File: src/main/java/at/xirado/bean/translation/LocaleLoader.java
package at.xirado.bean.translation;

import at.xirado.bean.misc.Util;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.interactions.DiscordLocale;
import net.dv8tion.jda.api.utils.data.DataObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class LocaleLoader {

    private static final Logger log = LoggerFactory.getLogger(LocaleLoader.class);

    public static final List<String> LANGUAGES = new ArrayList<>();
    private static final Map<String, DataObject> LANGUAGE_MAP;

    static {
        Map<String, DataObject> m = new HashMap<>();

        try (var is = LocaleLoader.class.getResourceAsStream("/assets/languages/list.txt")) {
            if (is == null) {
                throw new ExceptionInInitializerError("Could not initialize Language loader because list.txt does not exist!");
            }
            for (var lang : new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8)).lines().toList()) {
                var language = lang.trim();
                LANGUAGES.add(language);
            }
        } catch (IOException e) {
            log.error("Could not initialize locale loader!");
            throw new ExceptionInInitializerError(e);
        }

        for (String lang : LANGUAGES) {
            try (var is = LocaleLoader.class.getResourceAsStream("/assets/languages/" + lang)) {
                if (is == null) {
                    log.error("Locale {} does not exist!", lang);
                    continue;
                }
                var name = lang.replace(".json", "");
                DataObject json = DataObject.fromJson(is);
                m.put(name, json);
                log.info("Successfully loaded locale {}", lang);
            } catch (Exception e) {
                log.error("Could not load locale '{}'!", lang, e);
            }
        }

        LANGUAGE_MAP = Collections.unmodifiableMap(m);
    }

    public static DataObject getForLanguage(String language) {
        var lang = LANGUAGE_MAP.get(language);
        if (lang == null) {
            return LANGUAGE_MAP.get("en_US");
        }

        return lang;
    }

    public static DataObject ofGuild(Guild guild) {
        DiscordLocale locale = guild.getLocale();
        String tag = locale.getLocale();
        if (LANGUAGE_MAP.containsKey(tag)) {
            return LANGUAGE_MAP.get(tag);
        } else {
            return LANGUAGE_MAP.get("en_US");
        }
    }

    public static boolean isValidLanguage(String lang) {
        return LANGUAGE_MAP.containsKey(lang);
    }


    public static String parseDuration(long seconds, DataObject languageJSON, String delimiter) {
        if (seconds == -1) {
            return Util.getRecursive(languageJSON, "time.permanent");
        }
        long days = TimeUnit.SECONDS.toDays(seconds);
        long hours = TimeUnit.SECONDS.toHours(seconds) - (days * 24);
        long minutes = TimeUnit.SECONDS.toMinutes(seconds) - (TimeUnit.SECONDS.toHours(seconds) * 60);
        long seconds1 = TimeUnit.SECONDS.toSeconds(seconds) - (TimeUnit.SECONDS.toMinutes(seconds) * 60);
        StringBuilder ges = new StringBuilder();
        if (days != 0) {
            if (days == 1) {
                ges.append(days).append(" ").append(Util.getRecursive(languageJSON, "time.day")).append(delimiter);
            } else {
                ges.append(days).append(" ").append(Util.getRecursive(languageJSON, "time.days")).append(delimiter);
            }
        }
        if (hours != 0) {
            if (hours == 1) {
                ges.append(hours).append(" ").append(Util.getRecursive(languageJSON, "time.hour")).append(delimiter);
            } else {
                ges.append(hours).append(" ").append(Util.getRecursive(languageJSON, "time.hours")).append(delimiter);
            }
        }
        if (minutes != 0) {
            if (minutes == 1) {
                ges.append(minutes).append(" ").append(Util.getRecursive(languageJSON, "time.minute")).append(delimiter);
            } else {
                ges.append(minutes).append(" ").append(Util.getRecursive(languageJSON, "time.minutes")).append(delimiter);
            }
        }
        if (seconds1 != 0) {
            if (seconds1 == 1) {
                ges.append(seconds1).append(" ").append(Util.getRecursive(languageJSON, "time.second")).append(delimiter);
            } else {
                ges.append(seconds1).append(" ").append(Util.getRecursive(languageJSON, "time.seconds")).append(delimiter);
            }
        }
        String result = ges.toString();
        return result.substring(0, result.length() - delimiter.length());
    }
}

File: src/main/java/at/xirado/bean/translation/LocaleLoader.kt
package at.xirado.bean.translation

import at.xirado.bean.data.LinkedDataObject
import net.dv8tion.jda.api.entities.Guild
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeUnit

object LocaleLoader {
    private val logger = LoggerFactory.getLogger(this::class.java)
    private val languages = mutableListOf<String>()
    private val languageMap = mutableMapOf<String, LinkedDataObject>()

    init {
        this::class.java
            .getResourceAsStream("/assets/languages/list.txt")
            .use {
                val stream = it
                    ?: throw ExceptionInInitializerError("Could not initialize Language loader because list.txt does not exist!")
                stream.bufferedReader()
                    .readText()
                    .trim()
                    .lines()
                    .forEach(languages::add)
            }

        for (lang in languages) {
            this::class.java
                .getResourceAsStream("/assets/languages/$lang")
                .use {
                    val name = lang.removeSuffix(".json")
                    val json = LinkedDataObject.parse(it)
                    languageMap[name] = json.setMetadata(arrayOf(name))
                    logger.info("Successfully loaded locale $lang")
                }
        }
    }

    @JvmStatic
    fun getForLanguage(language: String) =
        languageMap[language] ?: languageMap.getValue("en_US")

    @JvmStatic
    fun ofGuild(guild: Guild) =
        getForLanguage(guild.locale.toLanguageTag())

    @JvmStatic
    fun parseDuration(seconds: Long, languageJSON: LinkedDataObject, delimiter: String): String {
        if (seconds == -1L) return languageJSON.getString("time.permanent")
        val days = TimeUnit.SECONDS.toDays(seconds)
        val hours = TimeUnit.SECONDS.toHours(seconds) - days * 24
        val minutes = TimeUnit.SECONDS.toMinutes(seconds) - TimeUnit.SECONDS.toHours(seconds) * 60
        val seconds1 = TimeUnit.SECONDS.toSeconds(seconds) - TimeUnit.SECONDS.toMinutes(seconds) * 60

        return buildString {
            if (days != 0L) {
                append("$days ")
                if (days == 1L) append(languageJSON.getString("time.day"))
                else append(languageJSON.getString("time.days"))
                append(delimiter)
            }

            if (hours != 0L) {
                append("$hours ")
                if (hours == 1L) append(languageJSON.getString("time.hour"))
                else append(languageJSON.getString("time.hours"))
                append(delimiter)
            }

            if (minutes != 0L) {
                append("$minutes ")
                if (minutes == 1L) append(languageJSON.getString("time.minute"))
                else append(languageJSON.getString("time.minutes"))
                append(delimiter)
            }

            if (seconds1 != 0L) {
                append("$seconds1 ")
                if (seconds1 == 1L) append(languageJSON.getString("time.second"))
                else append(languageJSON.getString("time.seconds"))
            }
        }.let { it.substring(0, it.length - delimiter.length) }
    }
}